README.md
# Bich
a wow addon aimimg to monitor mobs' ability


CHANGELOG.txt
2016-09-06 11:24
Change: Some bug fix.

2016-09-05 16:59
Change: Change toc file. Update the version field to 70000.
Add: Add some instance difficulty id.

2015-08-10 22:27
Change: Bugfixes.

2015-08-10 18:20
New Version: 0.5b
Change: Rewrite large amount of code.
Change: Change DB structure. (Attention! Update to this version will clear all local data!) 
Add: Add spellfilter.lua, which contains a spell filter aimimg to remove repeated spells in the future.
Add: Slash command \bichsearchdb & \bs are added to search spells by name/unitid/number of spells

2015-08-08 16:13
New Version: 0.4
Add: Shift + Left click will insert the spells to the chat input box.
Change: Change anchor of GameTooltip. It's near the spell icon now.


2015-08-08 15:34
New Version: 0.3
Change: Arrange spells in order of spellid. Will makes the spell bar more reading-friendly in some situations.


2015-08-08 14:10
New Version: 0.2
Change: Change some caption in the code.


2015-08-07 14:13
New Version: 0.1
First Release, Enjoy it!



Todo List
----
todo: add some useful functions to search in database
todo: filter enemies and friendly NPC
todo: add click behavior of spell to show ItemRefTooltip
todo: add CD tracker
todo: remove repeated spells (though they have different spellids)
todo: share spell information with others, API: SendAddonMessage()


bar.lua
--[[
only contains frontend (spell bar) functions
object: ns.BichBar
]]

local addon, ns = ...
local bcfg = ns.cfg.barConfig

local BichBar = CreateFrame("Frame", "BichSpellBar", UIParent)


function BichBar:init()
    self:SetWidth(200)
    self:SetHeight(50)
    self:SetPoint(bcfg.anchor, bcfg.relative.frame, bcfg.relative.anchor, bcfg.position.x, bcfg.position.y)
    self.buttons = {}
    self.buttonCount = 0
end


function BichBar:hideAllButtons()
    self.buttonCount = 0
    for i, button in ipairs(self.buttons) do
        button:Hide()
    end
end

function BichBar:createAndSetButton(spellid, i)
    local button = self:createButton(i)
    button:setButton(spellid)
    return button
end


function BichBar:createButton(i)
    local button = CreateFrame("Button", "BichSpellButton_"..tostring(i), self)
    local r, c = 0, i-1
    if bcfg.perRow ~= nil then
        r = math.floor((i-1)/bcfg.perRow)
        c = math.fmod(i-1,bcfg.perRow)
    end
    button:SetWidth(bcfg.button.width)
    button:SetHeight(bcfg.button.height)
    button:SetPoint("TOPLEFT", self, "TOPLEFT", 
                      (bcfg.button.width +bcfg.button.margin)*c,--(bcfg.button.width +bcfg.button.margin)*(c-(bcfg.perRow-1)/2),
                     -(bcfg.button.height+bcfg.button.margin)*r
                   )
    button.spellIcon = button:CreateTexture(nil, "ARTWORK")
    button.spellIcon:SetAllPoints(button)
    
    function button:setButton(spellid)
        local _, _, image = GetSpellInfo(spellid)
        self.spellIcon:SetTexture(image)
        --self.hyperlink = GetSpellLink(spellid)
        self.spellid = spellid
        self:Show()
    end
    
    function button:onEnter()
        if bcfg.tooltipAnchor == "ANCHOR_NONE" then
            GameTooltip_SetDefaultAnchor(GameTooltip, self)
        else
            GameTooltip:SetOwner(self, bcfg.tooltipAnchor)
        end
        --GameTooltip:SetHyperlink(self.hyperlink)
        GameTooltip:SetSpellByID(self.spellid)
        GameTooltip:Show()
    end
    
    function button:onLeave()
        GameTooltip:FadeOut()
    end
    
    
    function button:onClick(buttonType)
        if IsShiftKeyDown() and buttonType == "LeftButton" and ChatFrame1EditBox:IsVisible() then
            ChatFrame1EditBox:Insert(GetSpellLink(self.spellid))
        end
    end
    
    button:SetScript("OnEnter", button.onEnter)
    button:SetScript("OnLeave", button.onLeave)
    button:SetScript("OnMouseDown", button.onClick)
    return button
end

function BichBar:addButton(spellid,i)
    i = i or self.buttonCount+1
    if i > #self.buttons then
        local button = self:createAndSetButton(spellid, i)
        table.insert(self.buttons, button)
    else
        local button = self.buttons[i]
        button:setButton(spellid)
    end
end


function BichBar:createBar(spells)
    self:hideAllButtons()

    for i, spellid in ipairs(spells) do
        self.buttonCount = self.buttonCount + 1
        self:addButton(spellid, self.buttonCount)
    end
end


BichBar:init()
ns.BichBar = BichBar

core.lua
--[[
some core function. event handler, etc.
]] 

local addon, ns = ...
local gEnable = ns.cfg.enable
local fcfg = ns.cfg.filterConfig
local Bich = ns.Bich
local BichBar = ns.BichBar

-- some functions
local function printMobInfoToChat(mobInfo)
    local name = Bich:getUnitName(mobInfo.unitid)
    local msg = ""
    for i, spellid in ipairs(mobInfo.spells) do
        msg = msg .. GetSpellLink(spellid)
    end
    
    DEFAULT_CHAT_FRAME:AddMessage(name .. ":" .. msg, 255, 255, 255)
end


local function printSearchResultToChat(resultSet, limit)
    limit = limit or #resultSet -- no limit by default
    for i, mobInfo in ipairs(resultSet) do
        if i <= limit then
            printMobInfoToChat(mobInfo)
        end
    end
end

local function searchAndShow(cmd, param)
    searchResult = Bich:search(cmd, param)
    if #searchResult > 0 then
        local randomChoosen = searchResult[ fastrandom(1,#searchResult) ]
        BichBar:createBar(randomChoosen.spells)
        printSearchResultToChat(searchResult)
        --printSearchResultToChat({randomChoosen})
    end
end




-- initial
local function initAddon()
    Bich:checkVersion() -- check database version
    ns.loaded = true
    DEFAULT_CHAT_FRAME:AddMessage("|cff00f0f0Bich|r loaded", 255, 255, 255)
end

local function handlerAddonLoaded(name)
    if name == "Bich" then
        initAddon()
    end
end

-- get the spells and create bar
local function unitCreateBar(flag)
    local guid = UnitGUID(flag)
    local unitid = Bich:getCreatureIdByGuid(guid)
    local zone = Bich:getZone()
    local spells = Bich:getFilteredSpellInfo(zone, unitid, "sort")
    
    if next(spells) ~= nil then
        local name = UnitName(flag)
        Bich:setUnitName(unitid, name)
        BichBar:createBar(spells)
        --  for debug
        
        -- printMobInfoToChat({zone=zone, unitid=unitid, spells=spells})
    end
end

-- debug function
local function m_GetSpellLink(l)
    if GetSpellLink(l) then
        return GetSpellLink(l)
    else
        return l
    end
end

local function m_legalString(str)
    if type(str)  == "string" then
        return str
    elseif type(str) == "number" then
        return string.format("[%d]", str)
    elseif type(str) == "nil" then
        return "nil"
    else
        return type(str)
    end
end
    
-- store spell information
local function handlerCombatLog(timeStamp, event, hideCaster, sourceGUID, sourceName, sourceFlags, 
                                sourceRaidFlags, destGUID, destName, destFlags, destRaidFlags, ...)

    if not Bich:checkZone(fcfg.track) then
        return
    end

    -- if select(1,...) ~= nil and not string.find(event, "^SWING") then   -- "SWING_xxx" events dont have spell id.
        -- local l = select(1,...)
        -- DEFAULT_CHAT_FRAME:AddMessage(event..":"..m_legalString(sourceName)..":"..m_GetSpellLink(l), 255, 255, 255)
    -- end

    if event == "SPELL_HEAL" or event == "SPELL_CAST_START" or event == "SPELL_CAST_SUCCESS" or event == "SPELL_PERIODIC_DAMAGE" or event == "SPELL_MISSED" then
        local spellid = select(1, ...)
        local unitid = Bich:getCreatureIdByGuid(sourceGUID)
        local zone, zoneName = Bich:getZone()
        if unitid > 0 then
            local newSpell = Bich:addSpellAndSetName(zone, zoneName, unitid, sourceName, spellid)
            local tarGuid = UnitGUID("target")
            if unitid == Bich:getCreatureIdByGuid(tarGuid) and newSpell then
                unitCreateBar("target")
            end
        end
    end
end


-- show spells stored when target changed
local function handlerTargetChanged()
    BichBar:hideAllButtons()
    if UnitGUID("target") == nil or not Bich:checkZone(fcfg.show) then   --target is nil or not shown in this zone
        return
    end
    unitCreateBar("target")
end

-- save BichDB after logout. with filter
local function handlerPlayerLogout()
    for zone, _ in pairs(BichDB) do
        if not Bich:checkZone(fcfg.save, zone) then
            BichDB[zone] = nil
        end
    end
    for zone, _ in pairs(BichAddition.zone) do
        if not Bich:checkZone(fcfg.save, zone) then
            BichAddition.zone[zone] = nil
        end
    end
end

--event handler
local frame = CreateFrame("Frame")

function frame:onEvent(event, ...)
    if gEnable then
        if ns.loaded then
            if event == "COMBAT_LOG_EVENT_UNFILTERED" then
                handlerCombatLog(...)
            elseif event == "PLAYER_TARGET_CHANGED" or event == "ZONE_CHANGED_NEW_AREA" then
                handlerTargetChanged()       
            elseif event == "PLAYER_LOGOUT" then
                handlerPlayerLogout()
            end
        elseif event == "ADDON_LOADED" then
            handlerAddonLoaded(...)
        end
    end
end

frame:RegisterEvent("COMBAT_LOG_EVENT_UNFILTERED")
frame:RegisterEvent("PLAYER_TARGET_CHANGED")
frame:RegisterEvent("ADDON_LOADED")
frame:RegisterEvent("PLAYER_LOGOUT")
frame:RegisterEvent("ZONE_CHANGED_NEW_AREA")
frame:SetScript("OnEvent", frame.onEvent)


-- slash commands
SLASH_BICH_RELOADER1 = "/rl"
SlashCmdList.BICH_RELOADER = function() ReloadUI() end

SLASH_BICH_CLEARDB1 = "/bichcleardb"
SLASH_BICH_CLEARDB2 = "/bclr"
SlashCmdList.BICH_CLEARDB = function()
    _G["BichDB"] = {}
    _G["BichAddition"].zone = {}
    _G["BichAddition"].unit = {}
end

SLASH_BICH_SEARCHDB1 = "/bichsearchdb"
SLASH_BICH_SEARCHDB2 = "/bs"

SlashCmdList.BICH_SEARCHDB = function(msg, editbox)
    local cmd, param = msg:match("^(%S*)%s*(.-)$")
    cmd = string.lower(cmd)
    if cmd == "num" or cmd == "number" then
        param = tonumber(param)
    elseif cmd == "mob" or cmd == "creature" or cmd == "unit" then
        param = tonumber(param)
    elseif cmd == "name" then
        -- do nothing
    else    -- invalid cmd
        param = nil
    end
    if param ~= nil then
        searchAndShow(cmd, param)
    else
        DEFAULT_CHAT_FRAME:AddMessage("Wrong parameters!", 255, 255, 255)
        DEFAULT_CHAT_FRAME:AddMessage("|cff00f0f0/bs|r num|unit|name  param", 255, 255, 255)
    end
end

class.lua
-- class(table) Bich
local addon, ns = ...

BichDB = BichDB or {}

local Bich = {}

function Bich:getCreatureIdByGuid(guid)
    local utype, _, server, instance, zone, id, suid = strsplit("-", guid)
    if utype == "Creature" then
        return tonumber(id)
    else
        return -1
    end
end

function Bich:getZone()
--GetCurrentMapAreaID()得到map area ID
--GetRealZoneText()得到真实名字
--GetZoneText()得到名字（在地城中会有问题）
--GetMapNameByID(mapId)从mapid得到地图名字
    return 0
end


function Bich:addSpell(zone, unitid, spellid)
    if BichDB[zone] == nil then
        BichDB[zone] = {}
    end
    if BichDB[zone][unitid] == nil then
        BichDB[zone][unitid] = {}
    end
    if BichDB[zone][unitid].spells == nil then
        BichDB[zone][unitid].spells = {}
    end
    BichDB[zone][unitid].spells[spellid] = true
end

function Bich:setName(zone, unitid, name)
    if BichDB[zone] == nil then
        BichDB[zone] = {}
    end
    if BichDB[zone][unitid] == nil then
        BichDB[zone][unitid] = {}
    end
    BichDB[zone][unitid].name = name
end

function Bich:getAllInfo(zone, unitid)
    return BichDB[zone] and BichDB[zone][unitid] or {}
end

ns.Bich = Bich

config.lua
--[[
config file
object: ns.cfg
]] 

local addon, ns = ...
local cfg = {}
local barConfig = {}
local filterConfig = {}

-- global enable
cfg.enable = true


-- spellbar config
-- spellbar position
barConfig.position = {
    x = 0,
    y = -180,
}

-- spellbar relatives
barConfig.relative = {
    frame = UIParent,
    anchor = "CENTER",
}

barConfig.anchor = "CENTER"

-- set nil to arrange every icon in one row
barConfig.perRow = nil

-- anchor of tooltip of spells. avialable value:
-- ANCHOR_NONE, ANCHOR_CURSOR, ANCHOR_BOTTOMLEFT, ANCHOR_BOTTOMRIGHT, ANCHOR_LEFT, ANCHOR_RIGHT, ANCHOR_TOPLEFT, ANCHOR_TOPRIGHT
barConfig.tooltipAnchor = "ANCHOR_CURSOR"

-- icon size
barConfig.button = {
    height = 36,
    width = 36,
    margin = 4,
}

-- filter config
-- where to track the mobs' abilities
filterConfig.track = {
    Dungeon = true,
    Raid = true,
    Scenario = true,
    None = true,
}

-- where to show the bar of the mobs' abilities
filterConfig.show = {
    Dungeon = nil,
    Raid = nil,
    Scenario = nil,
    None = nil,
}

-- where to save to local file when player logout
filterConfig.save = {
    Dungeon = nil,
    Raid = nil,
    Scenario = nil,
    None = nil,
}

-- if set nil, filterConfig.show and filterConfig.save will follow filterConfig.track
for k, v in pairs(filterConfig.track) do
    if filterConfig.show[k] == nil then
        filterConfig.show[k] = v
    end
    if filterConfig.save[k] == nil then
        filterConfig.save[k] = v
    end
end

-- do not touch~
cfg.barConfig = barConfig
cfg.filterConfig = filterConfig
ns.cfg = cfg

spellfilter.lua
--[[
Some spells will generate many similar spell effects, with similar spellid, same icon and same description. These spellid should be saved in database but not shown in spell bar. this file implements spellfilter to filter all spell icon in database, and show some of them.

ns.BichSpellFilter
]]

local addon, ns = ...

local BichSpellFilter = {}

function BichSpellFilter:raw(spells)
    spells = spells or {}
    local spellList = {}
    for spellid, _ in pairs(spells) do
        table.insert(spellList, spellid)
    end
    return spellList
end

function BichSpellFilter:sort(spells)
    local spellList = self:raw(spells)
    table.sort(spellList)
    return spellList
end

function BichSpellFilter:delNoCd(spells)   ---delete the spell with same icon and name.
    local spellTable = {}
    for k, v in spells do   -- copy
        spellTable[k] = v
    end
    for spellid, _ in pairs(spellTable) do
        local name, _, icon = GetSpellInfo(spellid)
        -- GetSpellCooldown
        for s2, _ in pairs(spellTable) do
            if spellid ~= s2 then
                local n2, _, i2 = GetSpellInfo(s2)
                if name == n2 and icon == i2 then
                    spellTable[s2] = nil
                end
            end
        end
    end
end

function BichSpellFilter:isSameSpell(spellid1, spellid2)
    local name1, _, icon1 = GetSpellInfo(spellid1)
    local name2, _, icon2 = GetSpellInfo(spellid2)
    if name1 == name2 and icon1 == icon2 then
        return true
    else
        return false
    end
end

ns.BichSpellFilter = BichSpellFilter


db.lua
--[[
only contains database functions
object: ns.Bich
data: BichDB, BichAddition
]]
local addon, ns = ...
local BichSpellFilter = ns.BichSpellFilter

BichDB = BichDB or {}       -- only store spell info
local BichDBVersion = 1     -- BichAddition save creature id, names, zone names, zone id
local Bich = {
    instanceType = {
        [0] = "None",
        [1] = "5-Player Normal",
        [2] = "5-Player Heroic",
        [3] = "10-Player Raid",
        [4] = "25-Player Raid",
        [5] = "10-Player Heroic Raid",
        [6] = "25-Player Heroic Raid",
        [7] = "25-Player Looking for Raid",
        [8] = "Challenge Mode",
        [9] = "40-Player Raid",
        [10] = "Not Used",
        [11] = "Heroic Scenario",
        [12] = "Normal Scenario",
        [13] = "Not Used",
        [14] = "10-30-Player Normal Raid",
        [15] = "10-30-Player Heroic Raid",
        [16] = "20-Player Mythic Raid",
        [17] = "10-30-Player Looking for Raid",
        [18] = "40-player Event raid",  -- Used by the level 100 version of Molten Core for WoW's 10th anniversary
        [19] = "5-player Event",  -- Used by the level 90 version of UBRS at WoD launch
        [20] = "Event Scenario",  -- unknown usage
        [23] = "5-Player Mythic",
        [24] = "5-Player Timewalker",
    },
    
    instanceCategory = {
        [0] = "None",
        [1] = "Dungeon",
        [2] = "Dungeon",
        [3] = "Raid",
        [4] = "Raid",
        [5] = "Raid",
        [6] = "Raid",
        [7] = "Raid",
        [8] = "Dungeon",
        [9] = "Raid",
        [11] = "Scenario",
        [12] = "Scenario",
        [14] = "Raid",
        [15] = "Raid",
        [16] = "Raid",
        [17] = "Raid",
        --[18] = "Event",
        --[19] = "Event",
        --[20] = "Event Scenario",
        [23] = "Dungeon",
    },
}

function Bich:checkVersion()
    if BichAddition == nil or BichAddition.version ~= BichDBVersion then
        BichDB = {}
        BichAddition = { 
            version = BichDBVersion,
            unit = {}, 
            zone = {},
        }
    end
end


function Bich:getCreatureIdByGuid(guid)
    if guid == nil then return -1 end
    local utype, _, server, instance, zone, id, suid = strsplit("-", guid)
    if utype == "Creature" or utype == "Vehicle" then
        return tonumber(id)
    else
        return -1
    end
end

function Bich:getZone()
    --local mapId = GetCurrentMapAreaID()
    local name, _, difficultyID, _, _, _, _, instanceMapID = GetInstanceInfo()
    local suffix = "0"  --outside
    if difficultyID ~= 0 then   --in instances
        if self.instanceType[difficultyID] ~= nil then
            suffix = tostring(difficultyID)
        else
            suffix = 'i' --illegal
        end
    end
    return tostring(instanceMapID) .. "_" .. suffix, name
    --return mapId and mapName
end

-- check if the mob in this zone should be saved
function Bich:checkZone(filter, zone)
    zone = zone or self:getZone()
    local difficultyID = tonumber(select(2,strsplit("_", zone)))
    local category = self.instanceCategory[difficultyID]
    if category == nil then
        return false
    elseif filter[category] == nil then
        return filter["None"] or false
    else
        return filter[category] or false
    end
end

function Bich:addSpellAndSetName(zone, zoneName, unitid, unitName, spellid)
    self:addSpell(zone, unitid, spellid)
    self:setZoneName(zone, zoneName)
    self:setUnitName(unitid, unitName)
end

function Bich:addSpell(zone, unitid, spellid)
    if BichDB[zone] == nil then
        BichDB[zone] = {}
    end
    if BichDB[zone][unitid] == nil then
        BichDB[zone][unitid] = {}
    end
    if BichDB[zone][unitid][spellid] then
        return false
    else
        BichDB[zone][unitid][spellid] = true
        return true
    end
end

function Bich:setZoneName(zone, name)   -- may be different when difficulty differs
    BichAddition.zone[zone] = name
end

function Bich:getZoneName(zone)
    return BichAddition.zone[zone] or "Not Found."
end

function Bich:setUnitName(unitid, name)
    BichAddition.unit[unitid] = name
end

function Bich:getUnitName(unitid)
    return BichAddition.unit[unitid] or "Not Found."
end

function Bich:getAllSpellInfo(zone, unitid)
    return BichDB[zone] and BichDB[zone][unitid] or {}
end

function Bich:getFilteredSpellInfo(zone, unitid, cmd)
    local rawInfo = BichDB[zone] and BichDB[zone][unitid] or {}
    if cmd == "raw" then
        return BichSpellFilter:raw(rawInfo)
    elseif cmd == "sort" then
        return BichSpellFilter:sort(rawInfo)
    else
        return BichSpellFilter:raw(rawInfo)
    end
end


-- search

function Bich:search(cmd, param)
    local condition = nil
    if cmd == "num" or cmd == "number" then
        condition = function (unitid, mobSpells, param)
            local spellNum = #BichSpellFilter:raw(mobSpells)
            return type(mobSpells) == "table" and spellNum > param
        end
    elseif cmd == "mob" or cmd == "creature" or cmd == "unit" then
        condition = function (unitid, mobSpells, param) return unitid == param end
    elseif cmd == "name" then
        condition = function (unitid, mobSpells, param) return self:getUnitName(unitid) == param end
    else
        condition = function (unitid, mobSpells, param) return false end
    end
    
    local result = {}
    for zone, zoneInfo in pairs(BichDB) do
        for unitid, mobSpells in pairs(zoneInfo) do
            if condition(unitid, mobSpells, param) then
                table.insert(result, {zone=zone, unitid=unitid, spells=BichSpellFilter:sort(mobSpells)})
            end
        end
    end
    return result
end
    
ns.Bich = Bich
